odoo/tools/populate.py

```diff
@@ -0,0 +1,84 @@
+import random
+from odoo.tools import pycompat
+
+
+# format str
+# compat {** ... }
+
+def Random(seed):
+    r = random.Random()
+    r.seed(seed, version=2)
+    return r
+
+
+def format_str(val, counter, values): # todo format
+    if isinstance(val, str):
+        return val.format(counter=counter, values=values)
+    return val
+
+
+def chain_generators(fields_generates, model_name):
+    generate = root_generate()
+    for (fname, field_generate) in fields_generates:
+        generate = field_generate(generate, fname, model_name)
+    return generate
+
+
+def root_generate():
+    yield {}, False
+    while True:
+        yield {}, True
+
+
+def randomize(vals, weights=None, seed=False, formater=format_str):
+    def generate(iterator, field_name, model_name, counter_offset=0):
+        r = Random(seed or '%s+field+%s' % (model_name, field_name))
+        for counter, (values, complete) in enumerate(iterator):
+            val = r.choices(vals, weights)[0]
+            values[field_name] = formater(val, counter+counter_offset, values)
+            yield values, complete
+    return generate
+
+
+def cartesian(vals, weights=None, seed=False, formater=format_str):
+    def generate(iterator, field_name, model_name):
+        counter = 0
+        for counter, (values, complete) in enumerate(iterator):
+            if complete:
+                break
+            for val in vals:
+                yield {**values, field_name: formater(val, counter, values)}, False
+        yield from randomize(vals, weights, seed, formater)(iterator, field_name, model_name, counter)
+    return generate
+
+
+def iterate(vals, weights=None, seed=False, formater=format_str):
+    def generate(iterator, field_name, model_name):
+        counter = 0
+        for val in vals: # iteratable order is important, shortest first
+            (values, _) = next(iterator)
+            values[field_name] = formater(val, counter, values)
+            yield values, False
+            counter += 1
+        yield from randomize(vals, weights, seed, formater)(iterator, field_name, model_name, counter)
+    return generate
+
+
+def constant(val, formater=format_str):
+    def generate(iterator, field_name, _):
+        counter = 0
+        for values, complete in iterator:
+            values[field_name] = formater(val, counter, values)
+            yield values, complete
+            counter += 1
+    return generate
+
+
+def compute(function, seed=None):
+    def generate(iterator, field_name, model_name):
+        r = Random(seed or '%s+%s' % (model_name, field_name))
+        for counter, (values, complete) in enumerate(iterator):
```

Comment:
It looks like every generator has its own `counter`.
Are they overall consistent?

Reply:
Inside a cartesian product, no. For generators after Cartesian, they should be. Any other suggestion to manage counter is welcome.